### PR TITLE
[OPIK-167] Add number of dataset items in dataset page

### PR DESCRIPTION
**Pull request by BorisTkachenko**
## Details
Add number of dataset items in dataset page to be shown on UI

## Issues
OPIK-167

## Testing
Added corresponding check in integration tests

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java [hunk 1]
```diff
@@ -207,6 +216,24 @@ INSERT INTO dataset_items (
                 ;
             """;
 
+    private static final String FIND_DATASET_ITEMS_SUMMARY_BY_DATASET_IDS = """
+                SELECT
+                    count(id) AS count,
+                    dataset_id
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: minor
```suggestion
                    dataset_id,
                    count(id) AS count
```

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java [hunk 2]
```diff
@@ -55,6 +56,14 @@ public interface DatasetItemDAO {
     Flux<DatasetItem> getItems(UUID datasetId, int limit, UUID lastRetrievedId);
 
     Mono<List<WorkspaceAndResourceId>> getDatasetItemWorkspace(Set<UUID> datasetItemIds);
+
+    Flux<DatasetItemSummary> findDatasetItemSummaryByDatasetIds(Collection<UUID> datasetIds);
+
+    record DatasetItemSummary(UUID datasetId, long datasetItemsCount) {
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Minor: I would move this one out the interface

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java [hunk 3]
```diff
@@ -584,6 +611,27 @@ public Mono<List<WorkspaceAndResourceId>> getDatasetItemWorkspace(@NonNull Set<U
         });
     }
 
+    @Override
+    public Flux<DatasetItemSummary> findDatasetItemSummaryByDatasetIds(Collection<UUID> datasetIds) {
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: This help us the the trace of the method in New Relic

```suggestion
     @WithSpan
    public Flux<DatasetItemSummary> findDatasetItemSummaryByDatasetIds(Collection<UUID> datasetIds) {
```

**Review by thiagohora — APPROVED** [review]

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 3 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Updated

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 2 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Updated

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 1 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: @thiagohora Maybe I'm missing something, could you clarify please. Seems it's exactly the same, just different order?
Anyway I updated it)

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java [hunk 4]
```diff
@@ -55,6 +56,8 @@ public interface DatasetItemDAO {
     Flux<DatasetItem> getItems(UUID datasetId, int limit, UUID lastRetrievedId);
 
     Mono<List<WorkspaceAndResourceId>> getDatasetItemWorkspace(Set<UUID> datasetItemIds);
+
+    Flux<DatasetItemSummary> findDatasetItemSummaryByDatasetIds(Collection<UUID> datasetIds);
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Really minor: datasetIds better as Set. They better be unique and we don't care about the provided order.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java [hunk 5]
```diff
@@ -584,6 +605,28 @@ public Mono<List<WorkspaceAndResourceId>> getDatasetItemWorkspace(@NonNull Set<U
...
+            var statement = connection.createStatement(FIND_DATASET_ITEMS_SUMMARY_BY_DATASET_IDS)
+                    .bind("dataset_ids", datasetIds);
+
+            Segment segment = startSegment("dataset_items", "Clickhouse", "find_dataset_item_summary_by_dataset_ids");
+
+            return makeFluxContextAware(bindWorkspaceIdToFlux(statement))
+                    .doFinally(signalType -> endSegment(segment))
+                    .flatMap(result -> result.map((row, rowMetadata) -> new DatasetItemSummary(
+                            row.get("dataset_id", UUID.class),
+                            row.get("count", Long.class))));
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Very minor: use builder (with tobuilder = true) when possible, over a constructor.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemSummary.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.comet.opik.domain;
+
+import java.util.UUID;
+
+public record DatasetItemSummary(UUID datasetId, long datasetItemsCount) {
+    public static DatasetItemSummary empty(UUID datasetId) {
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemSummary.java]: Very minor: annotate arguments with Lombok non-null where it makes sense.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/DatasetsResourceTest.java [hunk 7]
```diff
@@ -1945,7 +1948,7 @@ private void findAndAssertPage(Dataset.DatasetPage actualEntity, int expected, i
         assertThat(actualEntity.content())
                 .usingRecursiveFieldByFieldElementComparatorIgnoringFields("id", "createdAt", "lastUpdatedAt",
                         "createdBy", "lastUpdatedBy", "experimentCount", "mostRecentExperimentAt",
-                        "workspaceName")
+                        "workspaceName", "datasetItemsCount")
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/DatasetsResourceTest.java]: Minor: better reuse `DATASET_IGNORED_FIELDS`.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetService.java [hunk 8]
```diff
@@ -171,10 +172,18 @@ public Dataset findById(@NonNull UUID id) {
...
                 .collect(toMap(ExperimentSummary::datasetId, Function.identity()));
 
+        Map<UUID, DatasetItemSummary> datasetItemSummaryMap = datasetItemDAO
+                .findDatasetItemSummaryByDatasetIds(List.of(dataset.id()))
+                .contextWrite(ctx -> AsyncUtils.setRequestContext(ctx, requestContext))
+                .toStream()
+                .collect(toMap(DatasetItemSummary::datasetId, Function.identity()));
+
         var summary = experimentSummary.computeIfAbsent(dataset.id(), ExperimentSummary::empty);
+        var datasetItemSummary = datasetItemSummaryMap.computeIfAbsent(dataset.id(), DatasetItemSummary::empty);
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetService.java]: Minor: this is copied and pasted multiple times (it was already, you're just following the convention). It'd be nice to encapsulate it into smaller private aux methods.

**Review by andrescrz — APPROVED** [review]
This is good. Just minor comments about the conventions in the service. Mostly for a cleaner code, but with no actual consequences.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 1 above)
Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Yeah, just for readability as the aggregation is based on `dataset_id`

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 4 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Updated

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemSummary.java (hunk 6 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemSummary.java]: Done

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/DatasetsResourceTest.java (hunk 7 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/DatasetsResourceTest.java]: Done

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetService.java (hunk 8 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetService.java]: Done

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java (hunk 5 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemDAO.java]: Done

**Review by andrescrz — APPROVED** [review]

**Pull request merged**